clients/drdisasm/drdisasm.cpp

```diff
@@ -0,0 +1,158 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#define DR_FAST_IR 1
+#include "dr_api.h"
+#include "droption.h"
+#include <iostream>
+#include <sstream>
+
+namespace {
+
+// XXX i#1684: We want cross-arch decoding support so a single build can decode
+// AArchXX and x86.  For now, a separate build is needed.
+#ifdef X86_64
+droption_t<std::string> op_mode(DROPTION_SCOPE_FRONTEND, "mode", "x64",
+                                "Decodes using the specified mode: 'x64' or 'x86'.",
+                                "Decodes using the specified mode: 'x64' or 'x86'.");
+#elif defined(ARM)
+droption_t<std::string> op_mode(DROPTION_SCOPE_FRONTEND, "mode", "arm",
+                                "Decodes using the specified mode: 'arm' or 'thumb'.",
+                                "Decodes using the specified mode: 'arm' or 'thumb'.");
+#endif
+droption_t<bool> op_show_bytes(DROPTION_SCOPE_FRONTEND, "show_bytes", true,
+                               "Display the instruction encoding bytes.",
+                               "Display the instruction encoding bytes.");
+
+static bool
+parse_bytes(std::string token, std::vector<byte> &bytes)
+{
+    // Assume everything is hex.
+    // Assume consecutive values should be split and endianized.
+    uint64 entry;
+    std::stringstream stream;
+    stream << std::hex << token;
+    if (!(stream >> entry))
+        return false;
+    do {
+        bytes.push_back(entry & 0xff);
+        entry >>= 8;
+    } while (entry > 0);
+    return true;
+}
+};
+
+int
+main(int argc, const char *argv[])
+{
+    int last_index;
+    std::string parse_err;
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, argv, &parse_err,
+                                       &last_index)) {
+        std::cerr << "Usage error: " << parse_err << "\nUsage:\n " << argv[0]
+                  << " [options] <hexadecimal bytes to decode as args or stdin>\n"
+                  << "Options:\n"
+                  << droption_parser_t::usage_short(DROPTION_SCOPE_ALL);
+        return 1;
+    }
+
+    void *dcontext = GLOBAL_DCONTEXT;
+
+#if defined(X86_64) || defined(ARM)
```

Comment:
Nit: For consistent readability style, add a one-liner comment in order to describe this code like you did for the other following main steps. Something like "// Set ISA mode if supplied"

Reply:
Done.